### PR TITLE
I29 wavefront

### PR DESCRIPTION
**Pull request by StanJBrown**
Wavefront now runs in 5 seconds the attached 175x500 image. Demonstrate how to use a java linked list as a stack or queue in Matlab which might be useful to others. Added an environment of E5 third floor. Also split the wavefront and path finding in the wavefront into two separate functions as one may one to change start points to look at paths.  

image

### CODE REVIEW

File: matlab_simulation/08-planning/lib/image_to_binary_map.m [hunk 1]
```diff
@@ -0,0 +1,18 @@
+function [imout] = image_to_binary_map( input_image)
+%image_to_binary converts an input image to a binary image
```
Comment (chutsu) [inline matlab_simulation/08-planning/lib/image_to_binary_map.m]: Consider adding 1 space after the `%` for readability, and also for other comment lines

File: matlab_simulation/08-planning/lib/wavefront.m [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+%set unreached cells to a lower negtivive number for display purposes when
+%using imagesc to show the map
+wavefrontmap(wavefrontmap(:,:)==-10)= -100;
+
+path = shortest_wavefront_path(wavefrontmap, startpos);
+
+end
+
+
+%alternative way to do what was done above. All matrices in matlab are
```
Comment (chutsu) [inline matlab_simulation/08-planning/lib/wavefront.m]: Put this in a different file?

File: matlab_simulation/08-planning/lib/shortest_wavefront_path.m [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            if ((wavefrontmap(wavefrontpnts(i,1),wavefrontpnts(i,2)) < current_min))
+                best_point = [wavefrontpnts(i,1),wavefrontpnts(i,2)]; 
+                current_min = wavefrontmap(curpos(1), curpos(2));
+            end
+    end 
+    curpos = best_point;
+    stack.addLast([best_point(1), best_point(2)]);
+end
+
+path = zeros(stack.size(), 2);
```
Comment (chutsu) [inline matlab_simulation/08-planning/lib/shortest_wavefront_path.m:62]: Add a little comment telling developer what the following loop does

File: matlab_simulation/08-planning/lib/shortest_wavefront_path.m [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                   [-1, -1];
+                   [0, -1];
+                   [-1, -1];
+                   ];
+curpos = starting_position;
+best_point = [curpos(1), curpos(2)];
+if(wavefrontmap(curpos(1), curpos(2)) < 0)    
+    error('Error. The start point is located in an obstical')
+end
+while wavefrontmap(curpos(1), curpos(2)) ~= 0
```
Comment (chutsu) [inline matlab_simulation/08-planning/lib/shortest_wavefront_path.m:35]: Add a newline and a small comment about what the loop does

File: matlab_simulation/08-planning/lib/wavefront.m [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+%add a 0 padding around the current map to ensure the environment is closed.
+%also means you do not need to check for boundary conditions. 
+curmap = padarray(curmap,1,0);
+
+% make an output map that is initilized to all -10, which encodes unexplored cells;
+wavefrontmap = -10*ones(size(curmap));
+
+%adjust the value of endpos to agree with the new padding. 
+endpos = endpos + [1,1];
+
```
Comment (ghost) [inline matlab_simulation/08-planning/lib/wavefront.m]: Do we need to adjust the value for startpos as well to accommodate for the padding?

Comment (StanJBrown) [inline matlab_simulation/08-planning/lib/wavefront.m]: Yes you are correct. Nice catch.

**Pull request merged**